src/prop/sat_proof_manager.h

```diff
@@ -0,0 +1,559 @@
+                  TheoryProxy* proxy,
+                  context::UserContext* userContext,
+                  ProofNodeManager* pnm);
+
+  /** Marks the start of a resolution chain.
+   *
+   * The given clause, at the node level, is registered in d_resLinks with a
+   * null pivot, since this is the first link in the chain.
+   *
+   * @param start a SAT solver clause (list of literals) that will be the first
+   * clause in a resolution chain.
+   */
+  void startResChain(const Minisat::Clause& start);
+  /** Adds a resolution step with a clause
+   *
+   * The resolution step is added to d_resLinks with the clause, at the node
+   * level, and the literal, at the node level, as the pivot.
+   *
+   * @param clause the clause being resolved against
+   * @param lit the pivot of the resolution step
+   */
+  void addResolutionStep(const Minisat::Clause& clause, Minisat::Lit lit);
+  /** Adds a resolution step with a unit clause
+   *
+   * The resolution step is added to d_resLinks such that lit, at the node level, is the pivot and
+   * and the unit clause is ~lit, at the node level.
+   *
+   * If the literal is marked as redundant, then a step is *not* is added to
+   * d_resLinks. It is rather saved to d_redundandLits, whose components we will
+   * be handled in a special manner when the resolution chain is finished. This
+   * is because the steps corresponding to the removal of redundant literals
+   * have to be done in a specific order. See proccessRedundantLits below.
+   *
+   * @param lit the literal being resolved against
+   * @param redundant whether lit is redundant
+   */
+  void addResolutionStep(Minisat::Lit lit, bool redundant = false);
+  /** Ends resolution chain concluding a unit clause
+   *
+   * This and the version below both call the node version of this method,
+   * described further below, which actually does the necessary processing.
+   */
+  void endResChain(Minisat::Lit lit);
+  /** Ends resolution chain concluding a clause */
+  void endResChain(const Minisat::Clause& clause);
+  /** Build refutation proof starting from conflict clause
+   *
+   * This and the versions below call the node version of this method, described
+   * further below, which actually does the necessary processing.
+   *
+   * @param adding whether the conflict is coming from a freshly added clause
+   */
+  void finalizeProof(const Minisat::Clause& inConflict, bool adding = false);
+  /** Build refutation proof starting from conflict unit clause
+   *
+   * @param adding whether the conflict is coming from a freshly added clause
+   */
+  void finalizeProof(Minisat::Lit inConflict, bool adding = false);
+  /** As above, but uses the unit conflict clause saved in d_conflictLit. */
+  void finalizeProof();
+  /** Set the unit conflict clause d_conflictLit. */
+  void storeUnitConflict(Minisat::Lit inConflict);
+
+  /** Retrive the refutation proof
+   *
+   * If there is no chain for false in d_resChains, an assumption proof node is
```

Comment:
I assume there are some dependencies for when this method should be called (after a call to finalizeProof maybe?).

It would be helpful to have some basic usage details on the order of the calls to these methods.  In other words, it would be nice to have something like a reg exp that describes the intended use of this class, for each check-sat call, like roughly:

`registerSatAssumptions; ( beginResChain (addResolutionStep )* endResChain)*; finalizeProof; getProof`

Reply:
This is nice indeed, I'll add and clarify the dependencies in the methods doc as well.